### PR TITLE
Evolve state when executing `ParametrizedEvolution` gates with `DefaultQubitJax`

### PR DESCRIPTION
**Pull request by AlbertMitjans**
Use `odeint` solver inside `DefaultQubitJax` to evolve the state of the device when executing `ParametrizedEvolution` gates.

### CODE REVIEW

File: pennylane/devices/default_qubit_jax.py [hunk 1]
```diff
@@ -177,6 +181,22 @@ def capabilities(cls):
         capabilities.update(passthru_interface="jax")
         return capabilities
 
+    def _apply_parametrized_evolution(self, state: TensorLike, operation: ParametrizedEvolution):
+        if operation.data is None or operation.t is None:
+            raise ValueError(
+                "The parameters and the time window are required to execute a ParametrizedEvolution "
+                "You can update its values by calling the ParametrizedEvolution class: EV(params, t)."
```
Comment (AlbertMitjans) [inline pennylane/devices/default_qubit_jax.py:188]:
```suggestion
                "You can update these values by calling the ParametrizedEvolution class: EV(params, t)."
```

**Comment by AlbertMitjans** [comment]
[sc-33458]

File: pennylane/devices/default_qubit_jax.py [hunk 2]
```diff
@@ -177,6 +181,22 @@ def capabilities(cls):
...
+            )
+
+        state = jnp.ravel(state)
+
+        def fun(y, t):
+            """dy/dt = -i H(t) y"""
+            return -1j * qml.matrix(operation.H(operation.data, t=t), wire_order=self.wires) @ y
+
+        result = odeint(fun, state, operation.t, **operation.odeint_kwargs)
+        return self._reshape(result[-1], [2] * self.num_wires)
```
Comment (Qottmann) [inline pennylane/devices/default_qubit_jax.py:198]: Out of curiosity, how does that reshaping work?

File: pennylane/devices/default_qubit_jax.py [hunk 3]
```diff
@@ -177,6 +181,22 @@ def capabilities(cls):
...
         return capabilities
 
+    def _apply_parametrized_evolution(self, state: TensorLike, operation: ParametrizedEvolution):
+        if operation.data is None or operation.t is None:
+            raise ValueError(
+                "The parameters and the time window are required to execute a ParametrizedEvolution "
+                "You can update these values by calling the ParametrizedEvolution class: EV(params, t)."
+            )
+
+        state = jnp.ravel(state)
```
Comment (Qottmann) [inline pennylane/devices/default_qubit_jax.py:191]: Are you sure that you need to do that? ``odeint`` is internally ravelling the input state, performing odeint on the flattened array, and then returns the unravelled output in the original shape.

Is state a pytree or jax array? Then it should not be a problem

**Review by Qottmann — COMMENTED** [review]
Wow that was quick @AlbertMitjans ⚡ 

At the moment it is _always_ computing the state vector evolution. Is there a way for it to change the wires of the operation and then decide whether it is applying the unitary implementation or the state vector evolution? The cutoff should be at `2**2n < 2**N` where `n` is the subsystem size and `N` the total number of qubits of the device.

File: pennylane/devices/default_qubit_jax.py (hunk 2 above)
Comment (AlbertMitjans) [reply pennylane/devices/default_qubit_jax.py:198]: It calls `jnp.reshape`, and it reshapes the flatten state into shape `[2] * self.num_wires` (e.g. `[2, 2, 2]` if 3 wires).

File: pennylane/devices/default_qubit_jax.py (hunk 3 above)
Comment (AlbertMitjans) [reply pennylane/devices/default_qubit_jax.py:191]: Yep, I tried it out without flattening the state, and an error was raised here:
```python
def fun(y, t):
            """dy/dt = -i H(t) y"""
            return -1j * qml.matrix(operation.H(operation.data, t=t), wire_order=self.wires) @ y
```
Because `y` was not flattened.

Comment (AlbertMitjans) [reply pennylane/devices/default_qubit_jax.py:191]: And yes, `state` is a jax array.

Comment (AlbertMitjans) [reply pennylane/devices/default_qubit_jax.py]: To follow the device convention, I changed `jnp.ravel` by `self._flatten` (which calls `jnp.ravel`, thus the same thing XD)

**Comment by AlbertMitjans** [comment]
> Is there a way for it to change the wires of the operation and then decide whether it is applying the unitary implementation or the state vector evolution?

Done! 👍 

> It might be tricky as wires may become runtime values, so hard to inspect in a jit pipeline

Do we want wires to be runtime values? Will we ever use wires as the input of the QNode?

**Comment by Qottmann** [comment]
Amazing @AlbertMitjans 😍 
Beats the statevector simulator and old implementation in all benchmarks 🚅 
(qutip is using a different algorithm (scipy vode), so that is another story)

**Computing gradients with random pwc:**
image

**Computing forward pass with random pwc**
image


**Computing forward pass with constant Hamiltonian**
image

File: tests/devices/test_default_qubit_jax.py [hunk 4]
```diff
@@ -512,6 +513,42 @@ def circuit():
 
         circuit()  # Just don't crash.
 
+    def test_parametrized_evolution_state_vector(self, mocker):
```
Comment (Qottmann) [inline tests/devices/test_default_qubit_jax.py:516]: 👌

File: tests/devices/test_default_qubit_jax.py [hunk 5]
```diff
@@ -512,6 +513,42 @@ def circuit():
...
+        """Test that when executing a ParametrizedEvolution with ``num_wires >= device.num_wires/2``
+        the `_evolve_state_vector_under_parametrized_evolution` method is used."""
+        dev = qml.device("default.qubit.jax", wires=1, shots=1000)
+        H = ParametrizedHamiltonian([1], [qml.PauliX(0)])
+        spy = mocker.spy(dev, "_evolve_state_vector_under_parametrized_evolution")
+
+        @jax.jit
+        @qml.qnode(dev, interface="jax", diff_method=None)
```
Comment (Qottmann) [inline tests/devices/test_default_qubit_jax.py:524]: why diff method none?

**Review by Qottmann — COMMENTED** [review]
I think we should add one integration test for the matrix variant because all tests for ParametrizedEvolution satisfy `n=N`

Happy to approve then, but would recommend another neutral reviewer 👍

File: tests/devices/test_default_qubit_jax.py (hunk 5 above)
Comment (AlbertMitjans) [reply tests/devices/test_default_qubit_jax.py:524]: It probably got copied over. Removing it 👍

Comment (AlbertMitjans) [inline tests/devices/test_default_qubit_jax.py:524]:
```suggestion
        @qml.qnode(dev, interface="jax")
```

**Comment by AlbertMitjans** [comment]
> I think we should add one integration test for the matrix variant because all tests for ParametrizedEvolution satisfy n=N

Changed number of device wires of one test (such that 2*n < N). I would avoid adding an extra test because they take a lot of time to compute.

File: tests/devices/test_default_qubit_jax.py [hunk 6]
```diff
@@ -512,6 +513,42 @@ def circuit():
...
+
+        @jax.jit
+        @qml.qnode(dev, interface="jax")
+        def circuit():
+            qml.evolve(H)(params=[], t=np.pi / 2)  # corresponds to a PauliX gate
+            return qml.expval(qml.PauliZ(0))
+
+        res = circuit()
+        spy.assert_called_once()
+        assert qml.math.allclose(res, -1)
```
Comment (mlxd) [inline tests/devices/test_default_qubit_jax.py:531]: I'd argue against the use of only pi/2, pi, etc, and results amounting to -1, 0, 1 etc.

Can a set of floating point parameters be chosen for the input that are analytically determined for the outputs?

File: tests/devices/test_default_qubit_jax.py [hunk 7]
```diff
@@ -512,6 +513,42 @@ def circuit():
...
+        @jax.jit
+        @qml.qnode(dev, interface="jax")
+        def circuit():
+            qml.evolve(H)(params=[], t=np.pi / 2)  # corresponds to a PauliX gate
+            return qml.expval(qml.PauliZ(0))
+
+        res = circuit()
+        spy.assert_not_called()
+        spy2.assert_called_once()
+        assert qml.math.allclose(res, -1)
```
Comment (mlxd) [inline tests/devices/test_default_qubit_jax.py:550]: As above.

**Review by mlxd — CHANGES_REQUESTED** [review]
I don't see any impact with other devices. Thanks @AlbertMitjans 
Some query around additional test case parameters.

**Comment by Qottmann** [comment]
[sc-33592]

**Review by lillian542 — APPROVED** [review]
Looks good! Nice to see how much it sped things up :)

**Review by Qottmann — APPROVED** [review]
Also all good from my side, really great job @AlbertMitjans 🚅 
I have been using this the past day already, much faster 👌

**Review by mlxd — APPROVED** [review]
Nice work!
